### PR TITLE
WT-6526 Fix assertion failure when opening DB in readonly

### PR DESCRIPTION
**Pull request by raviprakashgiri29**

### CODE REVIEW

**Comment by sueloverso** [comment]
This is still a draft. I have a couple comments.
1. We don't want the commit the changes to `ex_access.c`. That is just for the purposes of the user error report. Adding a python test would be the preferred way to test this.
2. The `wt_remove_if_exists` is already a wrapper and named in an expressive way that we don't want to lose. Also not all callers want to do the same thing in the case. I think something like this diff is where we want to go:
```
diff --git a/src/meta/meta_turtle.c b/src/meta/meta_turtle.c
index 00427acae..fb34c647c 100644
--- a/src/meta/meta_turtle.c
+++ b/src/meta/meta_turtle.c
@@ -232,7 +232,12 @@ __wt_turtle_init(WT_SESSION_IMPL *session)
      * Discard any turtle setup file left-over from previous runs. This doesn't matter for
      * correctness, it's just cleaning up random files.
      */
-    WT_RET(__wt_remove_if_exists(session, WT_METADATA_TURTLE_SET, false));
+    if ((ret = __wt_remove_if_exists(session, WT_METADATA_TURTLE_SET, false)) != 0) {
+        /* If we're a readonly database, we can skip discarding the leftover file. */
+        if (F_ISSET(S2C(session), WT_CONN_READONLY) && ret == EACCES)
+            ret = 0;
+        WT_RET(ret);
+    }

     /*
      * If we found a corrupted turtle file, then delete it and create a new. We could die after
diff --git a/src/os_common/filename.c b/src/os_common/filename.c
index f91d93c1e..1cb31f706 100644
--- a/src/os_common/filename.c
+++ b/src/os_common/filename.c
@@ -81,8 +81,12 @@ __wt_remove_if_exists(WT_SESSION_IMPL *session, const char *name, bool durable)
     bool exist;

     WT_RET(__wt_fs_exist(session, name, &exist));
-    if (exist)
+    if (exist) {
+        /* Modifications are not allowed on a read only database. */
+        if (F_ISSET(S2C(session), WT_CONN_READONLY))
+            return (EACCES);
         WT_RET(__wt_fs_remove(session, name, durable));
+    }
     return (0);
 }
```

**Comment by raviprakashgiri29** [comment]
> This is still a draft. I have a couple comments.
> 
> 1. We don't want the commit the changes to `ex_access.c`. That is just for the purposes of the user error report. Adding a python test would be the preferred way to test this.
> 2. The `wt_remove_if_exists` is already a wrapper and named in an expressive way that we don't want to lose. Also not all callers want to do the same thing in the case. I think something like this diff is where we want to go:
> 
> ```
> diff --git a/src/meta/meta_turtle.c b/src/meta/meta_turtle.c
> index 00427acae..fb34c647c 100644
> --- a/src/meta/meta_turtle.c
> +++ b/src/meta/meta_turtle.c
> @@ -232,7 +232,12 @@ __wt_turtle_init(WT_SESSION_IMPL *session)
>       * Discard any turtle setup file left-over from previous runs. This doesn't matter for
>       * correctness, it's just cleaning up random files.
>       */
> -    WT_RET(__wt_remove_if_exists(session, WT_METADATA_TURTLE_SET, false));
> +    if ((ret = __wt_remove_if_exists(session, WT_METADATA_TURTLE_SET, false)) != 0) {
> +        /* If we're a readonly database, we can skip discarding the leftover file. */
> +        if (F_ISSET(S2C(session), WT_CONN_READONLY) && ret == EACCES)
> +            ret = 0;
> +        WT_RET(ret);
> +    }
> 
>      /*
>       * If we found a corrupted turtle file, then delete it and create a new. We could die after
> diff --git a/src/os_common/filename.c b/src/os_common/filename.c
> index f91d93c1e..1cb31f706 100644
> --- a/src/os_common/filename.c
> +++ b/src/os_common/filename.c
> @@ -81,8 +81,12 @@ __wt_remove_if_exists(WT_SESSION_IMPL *session, const char *name, bool durable)
>      bool exist;
> 
>      WT_RET(__wt_fs_exist(session, name, &exist));
> -    if (exist)
> +    if (exist) {
> +        /* Modifications are not allowed on a read only database. */
> +        if (F_ISSET(S2C(session), WT_CONN_READONLY))
> +            return (EACCES);
>          WT_RET(__wt_fs_remove(session, name, durable));
> +    }
>      return (0);
>  }
> ```

Thanks @sueloverso for reviewing the changes. The changes in `ex_access.c` were just for testing purpose, I reproduced the same issue in a python test and added it for the review. 

Now the changes are ready to be reviewed.

File: test/suite/test_readonly_config.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        cursor.close()
+        self.conn.close()
+
+        olddir="."
+        newdir="RESTART"
+
+        # Simulate a crash by copying to a new directory(RESTART).
+        copy_wiredtiger_home(olddir, newdir)
+
+        shutil.copy(newdir+'/WiredTiger.turtle', newdir+'/WiredTiger.turtle.set')
```
Comment (tammybailey) [inline test/suite/test_readonly_config.py:71]: Is manually copying this file something we expect users to do when opening a database as read-only following a crash? Or are you trying to specifically recreate the scenario in WT-6526? In either case, I think we need a comment explaining why you are doing this, and in the latter case, perhaps it makes sense to put the ticket number in the test name?

File: test/suite/test_readonly_config.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        # Open wiredtiger in new directory and in readonly mode.
+        conn = self.wiredtiger_open(newdir, "readonly")
+
+        #conn.close()
```
Comment (tammybailey) [inline test/suite/test_readonly_config.py:76]: This line should be uncommented or removed.

Comment (raviprakashgiri29) [reply test/suite/test_readonly_config.py]: Yes, I have uncommented this line.

File: test/suite/test_readonly_config.py (hunk 1 above)
Comment (raviprakashgiri29) [reply test/suite/test_readonly_config.py:72]: I copied the file to reproduce the issue in WT-6526, updated the comment.

File: test/suite/test_readonly_config.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    # Create a table.
+    uri = "table:test_readonly"
+
+    def test_readonly_config(self):
+        nrows = 100
+        ds = SimpleDataSet(self, self.uri, nrows, key_format="S", value_format='u')
+        ds.populate()
+        value = b"aaaaa" * 10
+
+        # Initially load huge data
```
Comment (sueloverso) [inline test/suite/test_readonly_config.py:48]: We don't need huge data nor a lot of rows. This loop can go away. We've already populated the table with `ds.populate()`.

File: test/suite/test_readonly_config.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        cursor = self.session.open_cursor(self.uri)
+        for i in range(1, 10000):
+            cursor.set_key(ds.key(nrows + i))
+            cursor.set_value(value)
+            self.assertEquals(cursor.insert(), 0)
+        cursor.close()
+        self.session.checkpoint()
+
+        # Search the inserted values.
```
Comment (sueloverso) [inline test/suite/test_readonly_config.py:57]: We don't need to search the values. It is not related to the test.

File: test/suite/test_readonly_config.py (hunk 1 above)
Comment (sueloverso) [reply test/suite/test_readonly_config.py:72]: I initially had a comment asking to change the name of the file. But @tammybailey 's comment makes me think that this test/file would be better called `test_bug024.py`. Reproducing a bug is exactly what this test is.

File: test/suite/test_readonly_config.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            cursor.set_key(ds.key(nrows + i))
+            self.assertEquals(cursor.search(), 0)
+        cursor.close()
+        self.conn.close()
+
+        olddir="."
+        newdir="RESTART"
+
+        # Simulate a crash by copying to a new directory(RESTART).
+        copy_wiredtiger_home(olddir, newdir)
```
Comment (sueloverso) [inline test/suite/test_readonly_config.py:69]: We don't need to copy the directory. Just copying the file manually in the original directory is sufficient.

File: test/suite/test_readonly_config.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+from helper import copy_wiredtiger_home
+import wiredtiger, wttest
+from wtdataset import SimpleDataSet
+import os, shutil
+
+# test_readonly_config.py
+# test to create a test database, close all the connections and re-open it with readonly mode.
```
Comment (sueloverso) [inline test/suite/test_readonly_config.py:35]: I suggest a different file name below. I think this comment can be elaborated:
```
# WT-6526: test that we can successfully open a readonly connection after it was stopped while
# the temporary turtle file existed. We simulate that by copying the turtle file to its temporary name
# and then opening the connection readonly. 
```

File: test/suite/test_readonly_config.py (hunk 3 above)
Comment (raviprakashgiri29) [reply test/suite/test_readonly_config.py]: @sueloverso , Removed irrelevant insertion and searching of records.

File: test/suite/test_readonly_config.py (hunk 4 above)
Comment (raviprakashgiri29) [reply test/suite/test_readonly_config.py]: Fixed

File: test/suite/test_readonly_config.py (hunk 5 above)
Comment (raviprakashgiri29) [reply test/suite/test_readonly_config.py]: Fixed

File: test/suite/test_readonly_config.py (hunk 6 above)
Comment (raviprakashgiri29) [reply test/suite/test_readonly_config.py]: Renamed the python file and updated the comments.

File: test/suite/test_readonly_config.py (hunk 1 above)
Comment (raviprakashgiri29) [reply test/suite/test_readonly_config.py]: Fixed

**Pull request merged**